Traces: Examples of invalid EDHOC messages

emanjon opened the pull request.

marco-tiloca-sics commented on draft-ietf-lake-traces.md:
```diff
@@ -2817,6 +2838,73 @@ OSCORE Master Salt after KeyUpdate (Raw Value) (8 bytes)
+Invalid number of elements in the CBOR sequence.
+
+~~~~~~~~
+Invalid message_1 (38 bytes)
```
Yes, I agree.

The only exception would be if this particular EAD item is defined to have a mandatory value, but for that we'd need such an EAD item to be specified and registered already.

emanjon replied:
Good, I removed it for now as it we already have a "wrong type" example.

malishav commented:
These are some detailed tests on parsing message_1 and validating the point. The question is why would we do this only for message_1 and not for e.g. message_2 as well? I agree that these are very useful and I already caught several bugs in my implementation that need to be fixed for these tests to pass.

emanjon commented:
>I agree that these are very useful and I already caught several bugs in my implementation that need to be fixed for these tests to pass.

That is a very good argument for why we should include this.

>The question is why would we do this only for message_1 and not for e.g. message_2 as well?

As a first start I just came up with as many different types of invalidity as possible, trying to not repeat things. My idea was that the implementer would need to make sure that the same type of invalidity is checked in other parts as well. That should probably be written out.

Yes, I think it is a good idea to have some invalid examples for message_2 as well. There we could have a "wrong number of items" message_2 as well as PLAINTEXT_2 with invalid ID_CRED_R.

emanjon commented:
@malishav I added some invalid message_2 and PLAINTEXT_2 and added some more text in the beginning of the section. Please check

emanjon commented:
> This message indicates cipher suite 6, which has X25519 as EDHOC key exchange algorithm (ECDH curve).
> 
> Since the error is about G_X not corresponding to a point on the P-256 curve, it is better for consistency to indicate cipher suite 2, 3 or 5 instead.
> 

Indeed. I just used the example in 4.1. message_1 (first time) and missed that 6 is not P-256. I'll change to 2

emanjon commented on draft-ietf-lake-traces.md:
```diff
@@ -2817,6 +2838,94 @@ OSCORE Master Salt after KeyUpdate (Raw Value) (8 bytes)
+Invalid message_1 (37 bytes)
+03 06 58 20 a0 4e 73 60 1d f5 44 a7 0b a7 ea 1e 57 03 0f 7d 4b 4e b7
+f6 73 92 4e 58 d5 4c a7 7a 5e 7d 4d 4a 0e
+~~~~~~~~
```
fixed

marco-tiloca-sics commented on draft-ietf-lake-traces.md:
```diff
@@ -2817,6 +2838,94 @@ OSCORE Master Salt after KeyUpdate (Raw Value) (8 bytes)
+~~~~~~~~
+Invalid  PLAINTEXT_2 (15 bytes)
+27 a1 04 42 32 10 48 fa 5e fa 2e bf 92 0b f3
+~~~~~~~~
```
(This comment was made on a commit, but I add it here too to be sure it's not missed)



What is the invalidity in the new example?

This seems the correct way to encode the credential identifier of type 'kid' with value 0x3210.

emanjon replied:
My understanding of Section 3.5.3.2. is that the CBOR map { 4 : kid_x } is always replaced by the byte string kid_x.

I.e. the correct encoding of this PLAINTEXT_2 is 
27 42 32 10 48 fa 5e fa 2e bf 92 0b f3

marco-tiloca-sics replied:
Ops, you're right :-)

Maybe it's useful if these examples shortly mention what their invalidity is, possibly pointing to the relevant section of -lake-edhoc

emanjon replied:
That is a good idea

emanjon replied:
Done

emanjon commented:
Added that implementations must not compose these examples according to EDHOC. Added reference to RFC 8949

Added more invalid examples concerning SUITES_I and ID_CRED_R.

Please review.

marco-tiloca-sics commented on draft-ietf-lake-traces.md:
```diff
@@ -2817,6 +2838,118 @@ OSCORE Master Salt after KeyUpdate (Raw Value) (8 bytes)
+~~~~~~~~
+
+Invalid 16-bit encoding 19 00 03 of METHOD = 3. Correct is the deterministic encoding 03 according to Section 1.4 of {{I-D.ietf-lake-edhoc}} and Section 4.2.1 of {{RFC8949}}.
+
```
I think it is better to refer to Section 3.1 of -lake-edhoc (instead of Section 1.4).

That's where deterministic encoding is mentioned as a property of the protocol (emphasis mine):

> All **EDHOC messages** are CBOR Sequences [[RFC8742](https://www.rfc-editor.org/info/rfc8742)], and **are deterministically encoded**.

while Section 1.4 is more about terminology and language for the document (emphasis mine):

> When referring to CBOR, this **specification** always **refers** to Deterministically Encoded CBOR as specified in Sections 4.2.1 and 4.2.2 of [[RFC8949](https://www.rfc-editor.org/info/rfc8949)].

emanjon replied:
Done. I will try to update the EDHOC specification during AUTH48 so that section 3.1 refers to 4.2.1 of [[RFC8949](https://www.rfc-editor.org/info/rfc8949)].

marco-tiloca-sics commented on draft-ietf-lake-traces.md:
```diff
@@ -2817,6 +2838,118 @@ OSCORE Master Salt after KeyUpdate (Raw Value) (8 bytes)
+~~~~~~~~
+
+Invalid encoding of the map ID_CRED_R in PLAINTEXT_2. Correct is the deterministic bytewise lexicographic order encoding according to Section 1.4 of {{I-D.ietf-lake-edhoc}} and Section 4.2.1 of {{RFC8949}}.
+
```
Instead of Section 1.4 of -lake-edhoc, I think it is better to refer to both Sections 3.1 (as mentioning deterministic encoding) and Section 3.5.3 (as talking about ID_CRED_X and its encoding).

Note that Section 3.5.3 is not *explicitly* saying that the CBOR map encoding ID_CRED_X has to be deterministically encoded; it is just silently inheriting that requirement from Section 3.1.

Is it worth adding an explicit reminder in Section 3.5.3 in -lake-edhoc? Such a reminder is in fact present in Section 3.5.2 when discussing authentication credentials instead.

emanjon replied:
I will suggst to add an explicit reminder in Section 3.5.3 during AUTH48

marco-tiloca-sics commented on draft-ietf-lake-traces.md:
```diff
@@ -2817,6 +2838,118 @@ OSCORE Master Salt after KeyUpdate (Raw Value) (8 bytes)
+~~~~~~~~
+
+Invalid array encoding 81 02 of SUITES_I = 2. Correct encoding is 02 according to Section 5.2.1 of {{I-D.ietf-lake-edhoc}}.
+
```
i think it is more precise to refer to Section 5.2.2 of -lake-edhoc; that's where the two variants for encoding SUITES_I are defined.

emanjon replied:
Can do

emanjon commented:
If we want to have heading I think they should describe the general type of error instead of the specific example.

emanjon commented:
Maybe some headings
- Invalid G_X (cddl correct)
- Invalid cddl
- Nor deterministic encoding

and then subbullets inside

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Göran Selander ***@***.***>
Sent: Monday, September 4, 2023 5:09 PM
To: lake-wg/edhoc ***@***.***>
Cc: John Mattsson ***@***.***>; Author ***@***.***>
Subject: Re: [lake-wg/edhoc] Examples of invalid EDHOC messages (PR #442)


If we want to have heading I think they should describe the general type of error instead of the specific example.

"Text string encoding of G_X" and "G_X without leading zeroes" are both examples of "Invalid encoding of G_X", would you prefer two sections with the same heading?

Similarly, "x-coordinate of G_X not in interval [0, p-1]" and "x-coordinate of G_X not corresponding to a point" are both examples of "Invalid x-coordinate of G_X"

Alternative, we could just enumerate the cases: "Invalid case 1", "Invalid case 2", ...

Related question, should we skip the sub-sections and itemize in some other way (e.g. bullets)?

—
Reply to this email directly, view it on GitHub<https://protect2.fireeye.com/v1/url?k=31323334-501cfaf3-313273af-454445554331-52c95170901fff23&q=1&e=c235fe80-03ff-4ff1-b9d3-2726d78df225&u=https%3A%2F%2Fgithub.com%2Flake-wg%2Fedhoc%2Fpull%2F442%23issuecomment-1705425104>, or unsubscribe<https://protect2.fireeye.com/v1/url?k=31323334-501cfaf3-313273af-454445554331-7ed3269f70c38866&q=1&e=c235fe80-03ff-4ff1-b9d3-2726d78df225&u=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAFYXKYWQTFW3BISLWW6SYB3XYXVK3ANCNFSM6AAAAAA4EUBBWU>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

gselander commented:
The third category - non-deterministic CBOR - is not necessarily invalid. See discussion in #446.

emanjon commented:
Added a curve25519 point (p) of low order which produces the all zero value
https://github.com/jedisct1/libsodium/blob/master/src/libsodium/crypto_scalarmult/curve25519/ref10/x25519_ref10.c#L17
